File: elixir/list-ops/lib/list_ops.ex
defmodule ListOps do
  # Please don't use any external modules (especially List or Enum) in your
  # implementation. The point of this exercise is to create these basic
  # functions yourself. You may use basic Kernel functions (like `Kernel.+/2`
  # for adding numbers), but please do not use Kernel functions for Lists like
  # `++`, `--`, `hd`, `tl`, `in`, and `length`.

  @spec count(list) :: non_neg_integer
  def count(l), do: count(l, 0)
  defp count([], acc), do: acc
  defp count([_head | tail], acc), do: count(tail, acc + 1)

  @spec reverse(list) :: list
  def reverse(l), do: reverse(l, [])
  defp reverse([], acc), do: acc
  defp reverse([head | tail], acc), do: reverse(tail, [head | acc])

  @spec map(list, (any -> any)) :: list
  def map(l, f), do: map(l, [], f)
  defp map([], acc, _f), do: reverse(acc)
  defp map([head | tail], acc, f), do: map(tail, [f.(head) | acc], f)

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter(l, f), do: filter(l, [], f)
  defp filter([], acc, _f), do: reverse(acc)

  defp filter([head | tail], acc, f) do
    if f.(head) do
      filter(tail, [head | acc], f)
    else
      filter(tail, acc, f)
    end
  end

  @type acc :: any
  @spec reduce(list, acc, (any, acc -> acc)) :: acc
  def reduce([], acc, _f), do: acc
  def reduce([head | tail], acc, f), do: reduce(tail, f.(head, acc), f)

  @spec append(list, list) :: list
  def append(a, b) do
    a
    |> reverse()
    |> reduce(b, &[&1 | &2])
  end

  @spec concat([[any]]) :: [any]
  def concat(ll) do
    ll
    |> reverse()
    |> reduce([], &append/2)
  end
end

File: typescript/list-ops/list-ops.ts
type FilterFunction<T> = (el: T) => boolean
type MapFunction<T> = (el: T) => T
type FoldFunction<T> = (acc: T, el: T) => T

export default class List<T = number> {
  readonly values: T[]

  constructor(values: T[] = []) {
    this.values = values
  }

  append(other: List<T>): List<T> {
    const values: T[] = this._append(this.values, other.values)
    return new List(values)
  }

  concat(listOfLists: List<List<T>>): List<T> {
    const values: T[] = this._concat(this.values, listOfLists.values)
    return new List(values)
  }

  filter(fun: FilterFunction<T>): List<T> {
    const values: T[] = this._filter(fun, [], this.values)
    return new List(values)
  }

  map(fun: MapFunction<T>): List<T> {
    const values: T[] = this._map(fun, [], this.values)
    return new List(values)
  }

  length(): number {
    return this._length(this.values)
  }

  foldl(fun: FoldFunction<T>, acc: T): T {
    return this._foldl(fun, acc, this.values)
  }

  foldr(fun: FoldFunction<T>, acc: T): T {
    return this._foldr(fun, acc, this.values)
  }

  reverse(): List<T> {
    const values: T[] = this._reverse([], this.values)
    return new List(values)
  }

  private _append(values1: T[], values2: T[]): T[] {
    const values1Empty: boolean = values1.length === 0
    const values2Empty: boolean = values2.length === 0

    if (values1Empty) {
      if (values2Empty) {
        return []
      }
      return values2
    } else if (values2Empty) {
      return values1
    } else {
      return [...values1, ...values2]
    }
  }

  private _concat(acc: T[], lists: List<T>[]): T[] {
    if (lists.length === 0) {
      return acc
    }

    const [head, ...tail]: List<T>[] = lists
    return this._concat([...acc, ...head.values], tail)
  }

  private _filter(fun: FilterFunction<T>, acc: T[], values: T[]): T[] {
    if (values.length === 0) {
      return acc
    }

    const [head, ...tail]: T[] = values
    if (fun(head)) {
      acc = [...acc, head]
    }

    return this._filter(fun, acc, tail)
  }

  private _map(fun: MapFunction<T>, acc: T[], values: T[]): T[] {
    if (values.length === 0) {
      return acc
    }

    const [head, ...tail]: T[] = values
    acc = [...acc, fun(head)]
    return this._map(fun, acc, tail)
  }

  private _length(values: T[]): number {
    if (values.length === 0) {
      return 0
    }

    const [_head, ...tail]: T[] = values
    return 1 + this._length(tail)
  }

  private _foldl(fun: FoldFunction<T>, acc: T, values: T[]): T {
    if (values.length === 0) {
      return acc
    }

    const [head, ...tail]: T[] = values
    acc = fun(acc, head)
    return this._foldl(fun, acc, tail)
  }

  private _foldr(fun: FoldFunction<T>, acc: T, values: T[]): T {
    if (values.length === 0) {
      return acc
    }

    const [head, ...tail]: T[] = values
    acc = this._foldr(fun, acc, tail)
    return fun(acc, head)
  }

  private _reverse(acc: T[], values: T[]): T[] {
    if (values.length === 0) {
      return acc
    }

    const [head, ...tail]: T[] = values
    return [...this._reverse(acc, tail), head]
  }
}
